apollo-federation/src/sources/connect/json_selection/methods/future/arithmetic.rs

```diff
@@ -0,0 +1,356 @@
+                        None,
+                        vec_push(
+                            errors,
+                            ApplyToError::new(
+                                format!(
+                                    "Method ->{} requires numeric arguments",
+                                    method_name.as_ref()
+                                ),
+                                input_path.to_vec(),
+                                arg.range(),
+                            ),
+                        ),
+                    );
+                }
+            }
+            (Some(JSON::Number(result)), errors)
+        } else {
+            (
+                None,
+                vec![ApplyToError::new(
+                    format!(
+                        "Method ->{} requires numeric arguments",
+                        method_name.as_ref()
+                    ),
+                    input_path.to_vec(),
+                    method_name.range(),
+                )],
+            )
+        }
+    } else {
+        (
+            None,
+            vec![ApplyToError::new(
+                format!(
+                    "Method ->{} requires at least one argument",
+                    method_name.as_ref()
+                ),
+                input_path.to_vec(),
+                method_name.range(),
+            )],
+        )
+    }
+}
+
+macro_rules! infix_math_op {
+    ($name:ident, $op:tt) => {
+        fn $name(a: &Number, b: &Number) -> Option<Number> {
+            if a.is_f64() || b.is_f64() {
+                Number::from_f64(a.as_f64().unwrap() $op b.as_f64().unwrap())
+            } else if let (Some(a_i64), Some(b_i64)) = (a.as_i64(), b.as_i64()) {
+                infix_math_op!(@int_case a_i64, b_i64, $op)
+            } else {
+                None
+            }
+        }
+    };
+    // This branching is because when we are dividing, two whole numbers could result in a non-whole result (E.g. 7/2 = 3.5)
+    // It is much more predictable and intuitive for this to return "3.5" instead of "3" (if we did int math) but we need
+    // the return type to always be the same for type checking to work. So, for division specifically, we will always return
+    // a float.
+    (@int_case $a:ident, $b:ident, /) => {{
+        Number::from_f64($a as f64 / $b as f64)
+    }};
+    (@int_case $a:ident, $b:ident, $op:tt) => {{
+        Some(Number::from($a $op $b))
+    }};
+}
+infix_math_op!(add_op, +);
+infix_math_op!(sub_op, -);
+infix_math_op!(mul_op, *);
+infix_math_op!(div_op, /);
+infix_math_op!(rem_op, %);
+
+#[allow(dead_code)] // method type-checking disabled until we add name resolution
+fn math_shape(
+    method_name: &WithRange<String>,
+    _method_args: Option<&MethodArgs>,
+    _input_shape: Shape,
+    _dollar_shape: Shape,
+    _named_var_shapes: &IndexMap<&str, Shape>,
+    source_id: &SourceId,
+) -> Shape {
+    Shape::error("TODO: math_shape", method_name.shape_location(source_id))
```

Comment:
Just acknowledging this `TODO` as something we should get back to before making the math methods public.